brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+        if pairwise:
+        
+            # Fast Fourier transform along time dimension of data
+            fft_data = fft(data, axis=0)
+
+            # Shift pos and neg frequencies symmetrically, to keep signal real
+            fft_data[pos_freq, :, :] *= np.exp(1j * phase_shifts)
+            fft_data[neg_freq, :, :] *= np.exp(-1j * phase_shifts)
+
+            # Inverse FFT to put data back in time domain for ISC
+            shifted_data = np.real(ifft(fft_data, axis=0))
+
+            # Compute null ISC on shifted data for pairwise approach
+            shifted_isc = isc(shifted_data, pairwise=True,
+                              summary_statistic=summary_statistic, verbose=False)
+        
+        # In leave-one-out, apply shift only to each left-out participant
+        elif not pairwise:
+            
+            # Roll subject axis in phaseshifts for loop
+            phase_shifts = np.rollaxis(phase_shifts, 2, 0)
+            
+            shifted_isc = []
+            for s, shift in enumerate(phase_shifts):
+                
+                # Apply FFT to left-out subject
+                fft_subject = fft(data[:, :, s], axis=0)
+                
+                # Shift pos and neg frequencies symmetrically, to keep signal real
+                fft_subject[pos_freq, :] *= np.exp(1j * shift)
+                fft_subject[neg_freq, :] *= np.exp(-1j * shift)
+
+                # Inverse FFT to put data back in time domain for ISC
+                shifted_subject = np.real(ifft(fft_subject, axis=0))
+
+                # Compute ISC of shifted left-out subject against mean of N-1 subjects
+                nonshifted_mean = np.mean(np.delete(data, s, 2), axis=2)
+                loo_isc = isc(np.dstack((shifted_subject, nonshifted_mean)), pairwise=False,
+                              summary_statistic=None, verbose=False)
+                shifted_isc.append(loo_isc)
+                
+            # Get summary statistics across left-out subjects
+            if summary_statistic == np.mean:
+                shifted_isc = np.tanh(np.mean(np.arctanh(np.dstack(shifted_isc)), axis=2))
+            elif summary_statistic == np.median:
+                shifted_isc = np.median(np.dstack(shifted_isc), axis=2)
+                
+        distribution.append(shifted_isc)
+        
+        # Update random state for next iteration
+        random_state = np.random.RandomState(prng.randint(0, 2**32 - 1))
+        
+    # Convert distribution to numpy array
+    distribution = np.vstack(distribution)
+    assert distribution.shape == (n_shifts, n_voxels)
+
+    # Get p-value for actual median from shifted distribution
+    p = ((np.sum(np.abs(distribution) >= np.abs(observed), axis=0) + 1) /
+          float((len(distribution) + 1)))[np.newaxis, :]
+    
+    if return_distribution:
+        return observed, p, distribution
+    elif not return_distribution:
+        return observed, p
```

Comment:
If `return_distribution` is `false`, I would still return `distribution` as `None` to make sure the output is always a list of 3 elements. Currently, the user needs to write slightly different code based on the value of `return_distribution`, which is inconvenient.